README
NAME
    Geo::UK::Postcode - Object and class methods for working with British
    postcodes.

SYNOPSIS
        # See Geo::UK::Postcode::Regex for parsing/matching postcodes

        use Geo::UK::Postcode;

        my $pc = Geo::UK::Postcode->new("wc1h9eb");

        $pc->raw;             # wc1h9eb - as entered
        $pc->as_string;       # WC1H 9EB - output in correct format
        "$pc";                # stringifies, same output as '->as_string'
        $pc->fixed_format;    # 8 characters, the incode always last three

        $pc->area;            # WC
        $pc->district;        # 1
        $pc->subdistrict;     # H
        $pc->sector;          # 9
        $pc->unit;            # EB

        $pc->outcode;         # WC1H
        $pc->incode;          # 9EB

        $pc->strict;          # true if matches strict regex
        $pc->valid;           # true if matches strict regex and has a valid outcode
        $pc->partial;         # true if postcode is for a district or sector only

        $pc->non_geographical;    # true if outcode is known to be
                                  # non-geographical

        # Sort Postcode objects:
        use Geo::UK::Postcode qw/ pc_sort /;

        my @sorted_pcs = sort pc_sort @unsorted_pcs;

DESCRIPTION
    An object to represent a British postcode.

    For matching and parsing postcodes in a non-OO manner (for form
    validation, for example), see Geo::UK::Postcode::Regex

    For geo-location (finding latitude and longitude) see "GEO-LOCATING
    POSTCODES".

    Currently undef development - feedback welcome. Basic API unlikely to
    change greatly, just more features/more postcodes supported - see "TODO"
    list.

METHODS
  raw
    Returns exact string that the object was constructed from.

  as_string
        $pc->as_string;

        # or:

        "$pc";

    Stringification of postcode object, returns postcode with a single space
    between outcode and incode.

  fixed_format
        my $fixed_format = $postcode->fixed_format;

    Returns the full postcode in a fixed length (8 character) format, with
    extra padding spaces inserted as necessary.

  area, district, subdistrict, sector, unit
    Return the corresponding part of the postcode, undef if not present.

  outcode
    The first half of the postcode, before the space - comprises of the area
    and district.

  incode
    The second half of the postcode, after the space - comprises of the
    sector and unit. Returns an empty string if not present.

  outward, inward
    Aliases for "outcode" and "incode".

  valid
        if ($pc->valid) {
            ...
        }

    Returns true if postcode has valid outcode and matches strict regex.

  partial
        if ($pc->partial) {
          ...
        }

    Returns true if postcode is not a full postcode, either a postcode
    district ( e . g . AB10 ) or postcode sector (e.g. AB10 1).

  strict
        if ($pc->strict) {
          ...
        }

    Returns true if postcode matches strict regex, meaning all characters
    are valid ( although postcode might not exist ) .

  non_geographical
        if ($pc->non_geographical) {
          ...
        }

    Returns true if the outcode is known to be non-geographical. Note that
    geographical outcodes may have non-geographical postcodes within them.

    (Non-geographical postcodes are used for PO Boxes, or organisations
    receiving large amounts of post).

  bfpo
        if ($pc->bfpo) {
            ...
        }

    Returns true if postcode is mapped to a BFPO number (British Forces Post
    Office).

  posttowns
        my (@posttowns) = $postcode->posttowns;

    Returns list of one or more posttowns that this postcode is assigned to.

EXPORTABLE
  pc_sort
        my @sorted_pcs = sort pc_sort @unsorted_pcs;

    Exportable sort function, sorts postcode objects in a useful manner. The
    sort is in the following order: area, district, subdistrict, sector,
    unit (ascending alphabetical or numerical order as appropriate).

GEO-LOCATING POSTCODES
    Postcodes can be geolocated by obtaining the Ordnance Survey
    'Code-Point' data (or the free 'Code-Point Open' data).

    For full details of using this class with Code-Point data, see:
    Geo::UK::Postcode::Manual::Geolocation.

SEE ALSO
    *   Geo::UK::Postcode::Regex

    *   Geo::Address::Mail::UK

    *   Geo::Postcode

AUTHOR
    Michael Jemmeson <mjemmeson@cpan.org>

COPYRIGHT
    Copyright 2014- Michael Jemmeson

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Geo-UK-Postcode

0.008  2014-03-12 17:51:35 GMT

  [Maintenance]
  - removing outdated dzil READMEs (replacing with README.md from Dist::Milla)

0.007  2014-03-12 17:45:54 GMT

  - Requirement on Geo::UK::Postcode::Regex 0.012 added
  - Bugfixing tests - inconsistant undef/empty string depending on ::Regex version
  
  [Maintenance]
  - switching to Dist::Milla, including Build.PL

0.006     2013-12-20 12:58:37 Europe/London

  - Tidy, test tweaks, and added Geo::UK::Postcode::Regex 0.005 prereq

0.005     2013-10-23 12:28:19 Europe/London

  - Split Geo::UK::Postcode::Regex into new package

0.004     2013-10-01 17:48:02 Europe/London

  - Valid regex added (tests against valid outcodes)
  - Bugfixing regexes, documentation, tests

0.003     2013-09-27 14:16:05 Europe/London

  - Bugfixing strict parsing regex.

0.002     2013-05-22 09:35:16 Europe/London

  - Added basic non-geographical checks and BX1-99 districts

0.001     2013-05-12 21:52:50 Europe/London

  - Initial release


lib/Geo/UK/Postcode.pm
package Geo::UK::Postcode;

use Moo;
use MooX::Aliases;

use base 'Exporter';
use Geo::UK::Postcode::Regex;

use overload '""' => "as_string";

# ABSTRACT: Object and class methods for working with British postcodes.

our $VERSION = '0.006'; # VERSION

our @EXPORT_OK = qw/ pc_sort /;


has raw => ( is => 'ro' );    # Str
has components => (
    is      => 'rwp',
    default => sub { {} },
);


around BUILDARGS => sub {
    my ( $orig, $class, $args ) = @_;

    unless ( ref $args ) {
        $args = { raw => $args };
    }

    return $class->$orig($args);
};

sub BUILD {
    my ($self) = @_;

    die "No raw postcode supplied" unless $self->raw;

    my $pc = uc $self->raw;

    my $parsed = Geo::UK::Postcode::Regex->parse( $pc, { partial => 1 } )
        or die sprintf "Unable to parse '%s' as a postcode", $self->raw ;

    $self->_set_components($parsed);
}


sub as_string { $_[0]->outcode . ' ' . $_[0]->incode }


sub fixed_format {
    sprintf( "%-4s %-3s", $_[0]->outcode, $_[0]->incode );
}


sub area        { shift->components->{area} }
sub district    { shift->components->{district} }
sub subdistrict { shift->components->{subdistrict} }
sub sector      { shift->components->{sector} }
sub unit        { shift->components->{unit} }


sub outcode {
    sprintf( "%s%s%s",
        $_[0]->area, $_[0]->district, $_[0]->subdistrict || '' );
}

sub incode {
    ( $_[0]->sector // '' ) . ( $_[0]->unit || '' );
}


alias outward => 'outcode';
alias inward  => 'incode';


sub valid {
    $_[0]->components->{valid_outcode} && $_[0]->components->{strict} ? 1 : 0;
}

sub partial {
    $_[0]->components->{partial} ? 1 : 0;
}

sub strict {
    $_[0]->components->{strict} ? 1 : 0;
}


sub non_geographical {
    $_[0]->components->{non_geographical} ? 1 : 0;
}


sub bfpo {
    $_[0]->outcode eq 'BF1' ? 1 : 0;
}


sub posttowns { Geo::UK::Postcode::Regex->posttowns( $_[0]->outcode ) }


sub pc_sort($$) {
           $_[0]->area cmp $_[1]->area
        || $_[0]->district <=> $_[1]->district
        || ( $_[0]->subdistrict || '' ) cmp( $_[1]->subdistrict || '' )
        || ( $_[0]->incode || '' ) cmp( $_[1]->incode || '' );
}


1;

__END__

=pod

=encoding utf-8

=head1 NAME

Geo::UK::Postcode - Object and class methods for working with British postcodes.

=head1 VERSION

version 0.006

=head1 SYNOPSIS

  # See Geo::UK::Postcode::Regex for parsing/matching postcodes

  use Geo::UK::Postcode;
  
  my $pc = Geo::UK::Postcode->new( "wc1h9eb" );
  
  $pc->raw;            # wc1h9eb - as entered
  $pc->as_string;      # WC1H 9EB - output in correct format
  "$pc";               # stringifies, same output as '->as_string'
  $pc->fixed_format;   # 8 characters, the incode always last three
  
  $pc->area;           # WC
  $pc->district;       # 1
  $pc->subdistrict;    # H
  $pc->sector;         # 9
  $pc->unit;           # EB
  
  $pc->outcode;        # WC1H
  $pc->incode;         # 9EB
  
  $pc->strict;     # true if matches strict regex
  $pc->valid;      # true if matches strict regex and has a valid outcode
  $pc->partial;    # true if postcode is for a district or sector only

  $pc->non_geographical;    # true if outcode is known to be
                            # non-geographical

  # Sort Postcode objects:
  use Geo::UK::Postcode qw/ pc_sort /;
  
  my @sorted_pcs = sort pc_sort @unsorted_pcs;

=head1 DESCRIPTION

An object to represent a British postcode.

For matching and parsing postcodes in a non-OO manner (for form validation, for
example), see L<Geo::UK::Postcode::Regex>

For geo-location (finding latitude and longitude) see L</"GEO-LOCATING POSTCODES">.

Currently undef development - feedback welcome. Basic API unlikely to change
greatly, just more features/more postcodes supported - see L</TODO> list.

=for Pod::Coverage BUILDARGS BUILD

=head1 METHODS

=head2 raw

Returns exact string that the object was constructed from.

=head2 as_string

  $pc->as_string;

  # or:

  "$pc";

Stringification of postcode object, returns postcode with a single space
between outcode and incode.

=head2 fixed_format

    my $fixed_format = $postcode->fixed_format;

Returns the full postcode in a fixed length (8 character) format, with extra
padding spaces inserted as necessary.

=head2 area, district, subdistrict, sector, unit

Return the corresponding part of the postcode, undef if not present.

=head2 outcode

The first half of the postcode, before the space - comprises of the area and
district.

=head2 incode

The second half of the postcode, after the space - comprises of the sector
and unit. Returns an empty string if not present.

=head2 outward, inward

Aliases for C<outcode> and C<incode>.

=head2 valid

  if ($pc->valid) {
    ...
  }

Returns true if postcode has valid outcode and matches strict regex.

=head2 partial

  if ($pc->partial) {
    ...
  }

Returns true if postcode is not a full postcode, either a postcode district
( e . g . AB10 )
or postcode sector (e.g. AB10 1).

=head2 strict

  if ($pc->strict) {
    ...
  }

Returns true if postcode matches strict regex, meaning all characters are valid
( although postcode might not exist ) .

=head2 non_geographical

    if ($pc->non_geographical) {
      ...
    }

Returns true if the outcode is known to be non-geographical. Note that
geographical outcodes may have non-geographical postcodes within them.

(Non-geographical postcodes are used for PO Boxes, or organisations
receiving large amounts of post).

=head2 bfpo

    if ($pc->bfpo) {
        ...
    }

Returns true if postcode is mapped to a BFPO number (British Forces Post
Office).

=head2 posttowns

    my (@posttowns) = $postcode->posttowns;

Returns list of one or more posttowns that this postcode is assigned to.

=head1 EXPORTABLE

=head2 pc_sort

    my @sorted_pcs = sort pc_sort @unsorted_pcs;

Exportable sort function, sorts postcode objects in a useful manner. The
sort is in the following order: area, district, subdistrict, sector, unit
(ascending alphabetical or numerical order as appropriate).

=head1 GEO-LOCATING POSTCODES

Postcodes can be geolocated by obtaining the Ordnance Survey 'Code-Point' data
(or the free 'Code-Point Open' data).

For full details of using this class with Code-Point data, see:
L<Geo::UK::Postcode::Manual::Geolocation>.

=head1 SEE ALSO

=over

=item *

L<Geo::UK::Postcode::Regex>

=item *

L<Geo::Address::Mail::UK>

=item *

L<Geo::Postcode>

=back

=head1 TODO

=over

=item Finalise API

=item Handle non-geographic postcodes

=item Handle British overseas territories

=item Handle special case postcodes, like GIR 0AA and SAN TA1

=item Find out which BX non-geographical districts are used.

=item handle BFPO

=item cleanup method to correct common errors

=back

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/mjemmeson/geo-uk-postcode/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/mjemmeson/geo-uk-postcode>

  git clone git://github.com/mjemmeson/geo-uk-postcode.git

=head1 AUTHOR

Michael Jemmeson <mjemmeson@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Michael Jemmeson <mjemmeson@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Object and class methods for working with British postcodes.",
  "AUTHOR" => "Michael Jemmeson <mjemmeson\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Geo-UK-Postcode",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Geo::UK::Postcode",
  "PREREQ_PM" => {
    "Exporter" => 0,
    "Geo::UK::Postcode::Regex" => "0.005",
    "Moo" => 0,
    "MooX::Aliases" => 0,
    "base" => 0,
    "overload" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Find" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "List::Util" => 0,
    "Test::More" => 0,
    "Test::Most" => 0,
    "lib" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.006",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Find" => 0,
  "File::Spec::Functions" => 0,
  "File::Temp" => 0,
  "Geo::UK::Postcode::Regex" => "0.005",
  "List::Util" => 0,
  "Moo" => 0,
  "MooX::Aliases" => 0,
  "Test::More" => 0,
  "Test::Most" => 0,
  "base" => 0,
  "lib" => 0,
  "overload" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



